File: algorithms/2_medium/S/01927_sum_game/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static bool sumGame(const string& num) {
        const size_t len = num.size();
        assert((len & 1) == 0 && len >= 2);
        const size_t half = len / 2;
        isize balance_sum = 0;
        isize balance_cnt = 0;

        for (size_t i = 0; i < half; i += 1) {
            switch (num[i]) {
                case '?':
                    balance_cnt += 1;
                    break;
                default:
                    balance_sum += num[i] - '0';
                    break;
            }

            switch (num[i + half]) {
                case '?':
                    balance_cnt -= 1;
                    break;
                default:
                    balance_sum -= num[i + half] - '0';
                    break;
            }
        }

        if (balance_cnt == 0) {
            return balance_sum != 0;
        }

        if ((balance_cnt & 1) == 1 || balance_cnt * balance_sum >= 0 || (abs(balance_sum) % 9) != 0) {
            return true;
        }

        return (abs(balance_sum) / 9) != (abs(balance_cnt) / 2);
    }
};

File: algorithms/2_medium/S/01927_sum_game/solution.rs
impl Solution {
    pub fn sum_game(num: String) -> bool {
        let bytes = num.as_bytes();
        let len = bytes.len();
        assert!((len & 1) == 0 && len >= 2);
        let half = len / 2;
        let (mut balance_sum, mut balance_cnt) = (0_isize, 0_isize);

        for i in 0..half {
            match bytes[i] {
                b'?' => balance_cnt += 1,
                b'0'..=b'9' => balance_sum += (bytes[i] - b'0') as isize,
                _ => unreachable!()
            }

            match bytes[i + half] {
                b'?' => balance_cnt -= 1,
                b'0'..=b'9' => balance_sum -= (bytes[i + half] - b'0') as isize,
                _ => unreachable!()
            }
        }

        if balance_cnt == 0 {
            return balance_sum != 0;
        }

        if (balance_cnt & 1) == 1 || balance_cnt * balance_sum >= 0 || (balance_sum.abs() % 9) != 0 {
            return true;
        }

        (balance_sum.abs()) / 9 != (balance_cnt.abs() / 2)
    }
}
